Source/Core/InputCommon/ControllerInterface/Xlib/XInput2.cpp

```diff
@@ -196,13 +216,18 @@ KeyboardMouse::KeyboardMouse(Window window, int opcode, int pointer, int keyboar
     AddInput(new Cursor(!!(i & 2), !!(i & 1), (i & 2) ? &m_state.cursor.y : &m_state.cursor.x));
 
   // Mouse Axis, X-/+ and Y-/+
-  for (int i = 0; i != 4; ++i)
-    AddInput(new Axis(!!(i & 2), !!(i & 1), (i & 2) ? &m_state.axis.y : &m_state.axis.x));
+  for (int i = 0; i != 6; ++i)
+    AddInput(new Axis(i / 2, !!(i & 1),
+                      (i & 4) ? &m_state.axis.z :
+                      (i & 2) ? &m_state.axis.y :
+                                &m_state.axis.x));
 
   // Relative Mouse, X-/+ and Y-/+
   for (int i = 0; i != 4; ++i)
-    AddInput(new RelativeMouse(!!(i & 2), !!(i & 1),
-                               (i & 2) ? &m_state.relative_mouse.y : &m_state.relative_mouse.x));
+    AddInput(new RelativeMouse(i / 2, !!(i & 1),
+                               (i & 4) ? &m_state.relative_mouse.z :
+                               (i & 2) ? &m_state.relative_mouse.y :
+                                         &m_state.relative_mouse.x));
```

Comment:
There's some really ugly/confusing stuff going on here, please just unroll these loops so they're easier to understand.

Reply:
that's what I get for trying to stubbornly keep what was already there :P